cura/Machines/QualityManager.py

```diff
@@ -0,0 +1,474 @@
+
+        name = metadata["name"]
+        if name not in quality_type_node.children_map:
+            quality_type_node.children_map[name] = QualityChangesGroup(name, quality_type)
+        quality_changes_group = quality_type_node.children_map[name]
+        quality_changes_group.addNode(QualityNode(metadata))
+
+
+#
+# Similar to MaterialManager, QualityManager maintains a number of maps and trees for material lookup.
+# The models GUI and QML use are now only dependent on the QualityManager. That means as long as the data in
+# QualityManager gets updated correctly, the GUI models should be updated correctly too, and the same goes for GUI.
+#
+# For now, updating the lookup maps and trees here is very simple: we discard the old data completely and recreate them
+# again. This means the update is exactly the same as initialization. There are performance concerns about this approach
+# but so far the creation of the tables and maps is very fast and there is no noticeable slowness, we keep it like this
+# because it's simple.
+#
+class QualityManager(QObject):
+
+    qualitiesUpdated = pyqtSignal()
+
+    def __init__(self, container_registry, parent = None):
+        super().__init__(parent)
+        self._application = Application.getInstance()
+        self._material_manager = self._application._material_manager
+        self._container_registry = container_registry
+
+        self._empty_quality_container = self._application.empty_quality_container
+        self._empty_quality_changes_container = self._application.empty_quality_changes_container
+
+        self._machine_variant_material_quality_type_to_quality_dict = {}  # for quality lookup
+        self._machine_quality_type_to_quality_changes_dict = {}  # for quality_changes lookup
+
+        self._default_machine_definition_id = "fdmprinter"
+
+        self._container_registry.containerMetaDataChanged.connect(self._onContainerMetadataChanged)
+        self._container_registry.containerAdded.connect(self._onContainerMetadataChanged)
+        self._container_registry.containerRemoved.connect(self._onContainerMetadataChanged)
+
+        # When a custom quality gets added/imported, there can be more than one InstanceContainers. In those cases,
+        # we don't want to react on every container/metadata changed signal. The timer here is to buffer it a bit so
+        # we don't react too many time.
+        self._update_timer = QTimer(self)
+        self._update_timer.setInterval(300)
+        self._update_timer.setSingleShot(True)
+        self._update_timer.timeout.connect(self._updateMaps)
+
+    def initialize(self):
+        # Initialize the lookup tree for quality profiles with following structure:
+        # <machine> -> <variant> -> <material>
+        #           -> <material>
+
+        self._machine_variant_material_quality_type_to_quality_dict = {}  # for quality lookup
+        self._machine_quality_type_to_quality_changes_dict = {}  # for quality_changes lookup
```

Comment:
I have removed the outdated doc in aa49b0131. The lookup tree is indeed complex and this is because the data structures and lookup logics we have are this complex. The documentation in `initialize()` is correct about those two lookup trees.